File: app/src/main/java/com/memtrip/pinyin/audio/stream/Player.java
package com.memtrip.pinyin.audio.stream;

import android.content.Context;
import android.net.Uri;
import android.os.Handler;
import android.os.Looper;

import com.google.android.exoplayer2.DefaultLoadControl;
import com.google.android.exoplayer2.DefaultRenderersFactory;
import com.google.android.exoplayer2.ExoPlaybackException;
import com.google.android.exoplayer2.ExoPlayerFactory;
import com.google.android.exoplayer2.PlaybackParameters;
import com.google.android.exoplayer2.SimpleExoPlayer;
import com.google.android.exoplayer2.Timeline;
import com.google.android.exoplayer2.extractor.DefaultExtractorsFactory;
import com.google.android.exoplayer2.source.ExtractorMediaSource;
import com.google.android.exoplayer2.source.MediaSource;
import com.google.android.exoplayer2.source.TrackGroupArray;
import com.google.android.exoplayer2.trackselection.AdaptiveTrackSelection;
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector;
import com.google.android.exoplayer2.trackselection.TrackSelectionArray;
import com.google.android.exoplayer2.upstream.DefaultBandwidthMeter;
import com.google.android.exoplayer2.upstream.DefaultHttpDataSourceFactory;
import com.memtrip.pinyin.BuildConfig;

import java.io.IOException;

import static com.google.android.exoplayer2.Player.STATE_BUFFERING;
import static com.google.android.exoplayer2.Player.STATE_READY;

public class Player {

    private final SimpleExoPlayer player;
    private final MediaSource mediaSource;
    private final OnPlayerStateListener onPlayerStateListener;

    public Player(String streamUrl,
                  final OnPlayerStateListener onPlayerStateListener,
                  Context context) {

        this(
                ExoPlayerFactory.newSimpleInstance(
                        new DefaultRenderersFactory(context),
                        new DefaultTrackSelector(new AdaptiveTrackSelection.Factory(new DefaultBandwidthMeter())),
                        new DefaultLoadControl()),

                new ExtractorMediaSource(
                        Uri.parse(streamUrl),
                        new DefaultHttpDataSourceFactory(new UserAgent().get(), null),
                        new DefaultExtractorsFactory(),
                        new Handler(Looper.getMainLooper()),
                        new ExtractorMediaSource.EventListener() {
                            @Override
                            public void onLoadError(IOException error) {
                                onPlayerStateListener.onBufferingError(error);
                            }
                        }),
                onPlayerStateListener
        );
    }

    private Player(SimpleExoPlayer player,
                   MediaSource mediaSource,
                   OnPlayerStateListener onPlayerStateListener) {

        this.player = player;
        this.mediaSource = mediaSource;
        this.onPlayerStateListener = onPlayerStateListener;
    }

    void prepare() {

        player.addListener(new com.google.android.exoplayer2.Player.EventListener() {
            @Override
            public void onTimelineChanged(Timeline timeline, Object manifest) {

            }

            @Override
            public void onTracksChanged(TrackGroupArray trackGroups, TrackSelectionArray trackSelections) {

            }

            @Override
            public void onLoadingChanged(boolean isLoading) {
                if (isLoading) {
                    onPlayerStateListener.startBuffering();
                }
            }

            @Override
            public void onPlayerStateChanged(boolean playWhenReady, int playbackState) {
                if (player.getPlaybackState() != STATE_BUFFERING) {
                    if (player.getPlaybackState() == STATE_READY && playWhenReady) {
                        onPlayerStateListener.onPlay();
                    } else {
                        onPlayerStateListener.onCompleted();
                    }
                }
            }

            @Override
            public void onRepeatModeChanged(int repeatMode) {

            }

            @Override
            public void onPlayerError(ExoPlaybackException error) {
                onPlayerStateListener.onBufferingError(null);
            }

            @Override
            public void onPositionDiscontinuity() {

            }

            @Override
            public void onPlaybackParametersChanged(PlaybackParameters playbackParameters) {

            }
        });
        player.prepare(mediaSource);
        player.setPlayWhenReady(true);
    }

    void release() {
        player.stop();
        player.release();
    }

    void pause() {
        player.setPlayWhenReady(false);
    }

    private static class UserAgent {
        public String get() {
            return BuildConfig.APPLICATION_ID + "/" + BuildConfig.VERSION_NAME;
        }
    }
}

File: app/src/main/java/com/consistence/pinyin/audio/stream/Player.kt
package com.consistence.pinyin.audio.stream

import android.content.Context
import android.net.Uri
import android.os.Handler
import android.os.Looper
import com.consistence.pinyin.BuildConfig
import com.google.android.exoplayer2.*
import com.google.android.exoplayer2.Player
import com.google.android.exoplayer2.Player.STATE_BUFFERING
import com.google.android.exoplayer2.Player.STATE_READY
import com.google.android.exoplayer2.extractor.DefaultExtractorsFactory
import com.google.android.exoplayer2.source.ExtractorMediaSource
import com.google.android.exoplayer2.source.MediaSource
import com.google.android.exoplayer2.source.TrackGroupArray
import com.google.android.exoplayer2.trackselection.AdaptiveTrackSelection
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector
import com.google.android.exoplayer2.trackselection.TrackSelectionArray
import com.google.android.exoplayer2.upstream.DefaultBandwidthMeter
import com.google.android.exoplayer2.upstream.DefaultHttpDataSourceFactory

class Player constructor(
        streamUrl: String,
        private val onPlayerStateListener: OnPlayerStateListener,
        context: Context,
        private val player: SimpleExoPlayer = ExoPlayerFactory.newSimpleInstance(
                DefaultRenderersFactory(context),
                DefaultTrackSelector(AdaptiveTrackSelection.Factory(DefaultBandwidthMeter())),
                DefaultLoadControl()),
        private val mediaSource: MediaSource = ExtractorMediaSource(
                Uri.parse(streamUrl),
                DefaultHttpDataSourceFactory(
                        BuildConfig.APPLICATION_ID + "/" + BuildConfig.VERSION_NAME, null),
                DefaultExtractorsFactory(),
                Handler(Looper.getMainLooper()),
                ExtractorMediaSource.EventListener { onPlayerStateListener.onBufferingError() })) {

    internal fun prepare() {

        player.addListener(object : EventListenerAdapter {

            override fun onLoadingChanged(isLoading: Boolean) {
                if (isLoading) {
                    onPlayerStateListener.startBuffering()
                }
            }

            override fun onPlayerStateChanged(playWhenReady: Boolean, playbackState: Int) {
                if (player.playbackState != STATE_BUFFERING) {
                    if (player.playbackState == STATE_READY && playWhenReady) {
                        onPlayerStateListener.onPlay()
                    } else {
                        onPlayerStateListener.onCompleted()
                    }
                }
            }

            override fun onPlayerError(error: ExoPlaybackException?) {
                onPlayerStateListener.onBufferingError()
            }
        })

        player.prepare(mediaSource)
        player.playWhenReady = true
    }

    internal fun release() {
        player.stop()
        player.release()
    }

    internal fun pause() {
        player.playWhenReady = false
    }
}

interface EventListenerAdapter :  Player.EventListener {
    override fun onTimelineChanged(timeline: Timeline?, manifest: Any?) { }
    override fun onTracksChanged(trackGroups: TrackGroupArray?, trackSelections: TrackSelectionArray?) { }
    override fun onLoadingChanged(isLoading: Boolean) { }
    override fun onPlayerStateChanged(playWhenReady: Boolean, playbackState: Int) { }
    override fun onRepeatModeChanged(repeatMode: Int) { }
    override fun onPlayerError(error: ExoPlaybackException?) { }
    override fun onPositionDiscontinuity() { }
    override fun onPlaybackParametersChanged(playbackParameters: PlaybackParameters?) { }
}